`kitcn migrate up --prod` / `kitcn aggregate --prod` fail unauthenticated when CONVEX_DEPLOY_KEY is set via ambient env — regression still present after #212

## Summary

PR #212 fixed the subprocess-env wipe for **`kitcn deploy`** (and the migration + aggregate flows it chains internally), but the **standalone** `kitcn migrate` and `kitcn aggregate` subcommands were not patched. When invoked from CI with `CONVEX_DEPLOY_KEY` exported as an env var (the idiomatic GitHub Actions pattern), these commands still spawn the underlying `convex run` with the deployment keys stripped and immediately fail with exit code 1 against prod.

Same root cause as #208 / #209 in a different code path — not a new bug class.

**Affected version:** `0.13.2` (latest on npm as of 2026-04-18)

## Reproduction

```yaml
# .github/workflows/deploy.yml
- name: Run Convex migrations (prod only)
  working-directory: packages/backend
  env:
    CONVEX_DEPLOY_KEY: ${{ secrets.CONVEX_DEPLOY_KEY }}  # prod:<name>|<token>
  run: bunx kitcn migrate up --prod --yes
```

Result (CI log, 0.13.2):

```
##[group]Run bunx kitcn migrate up --prod --yes
  env:
    CONVEX_DEPLOY_KEY: ***
##[endgroup]
##[error]Process completed with exit code 1.
```

Exit happens in ~1 second with no stdout/stderr because `runMigrationFlow` calls `runBackendFunction(..., { echoOutput: false })`, so the underlying `convex run` auth failure is never surfaced.

Meanwhile, the preceding `bunx kitcn deploy --yes` step in the same job (same env, same deploy key) completes successfully, including its internal post-deploy migration check ("No pending migrations to apply.") and aggregate backfill. The contrast is the smoking gun: the env wipe affects the standalone subcommand path but not the deploy path, because only the deploy path was patched in #212.

## Root cause (0.13.2 `dist/cli.mjs` + `dist/backend-core-*.mjs`)

`handleDeployCommand` (cli.mjs ~L1759) does the right thing after #212:

```js
const deployCommandEnv = backend === "convex" ? getConvexDeploymentCommandEnv() : void 0;
// ...
await runMigrationFlow({
  execaFn, backendAdapter, migrationConfig, targetArgs,
  env: deployCommandEnv,           // ← env threaded in
  context: "deploy", direction: "up",
});
await runAggregateBackfillFlow({
  execaFn, backendAdapter, backfillConfig, mode: "resume",
  targetArgs, env: deployCommandEnv,  // ← same here
  context: "deploy",
});
```

`handleMigrateCommand` (cli.mjs ~L2322), the standalone path:

```js
if (migrateArgs.subcommand === "up") return runMigrationFlow({
  execaFn,
  backendAdapter,
  migrationConfig,
  targetArgs,
  context: "migration",
  direction: "up",
  // ← no `env` parameter, so runBackendFunction sees options.env === undefined
});
```

`handleAggregateCommand` (cli.mjs ~L1559), the standalone path:

```js
return runAggregateBackfillFlow({
  execaFn,
  backendAdapter,
  backfillConfig,
  mode: subcommand === "rebuild" ? "rebuild" : "resume",
  targetArgs,
  context: "aggregate",
  // ← same omission
});
```

Downstream, `runBackendFunction` (backend-core ~L15206) spawns the convex subprocess with:

```js
env: createBackendCommandEnv(options?.env),
```

and `createBackendCommandEnv` (backend-core ~L12889) unconditionally strips all four `CONVEX_DEPLOYMENT_ENV_KEYS`:

```js
function createBackendCommandEnv(overrides) {
  const clearedDeploymentEnv = Object.fromEntries(
    CONVEX_DEPLOYMENT_ENV_KEYS.map((key) => [key, void 0]),
  );
  return { ...createCommandEnv(), ...clearedDeploymentEnv, ...overrides };
}
```

Without `overrides` carrying the keys back in (which is exactly what `handleDeployCommand` now does post-#212), the subprocess runs with no `CONVEX_DEPLOY_KEY`, no `CONVEX_DEPLOYMENT`, no self-hosted vars. `--prod` is forwarded correctly by `extractConvexRunTargetArgs`, but `convex run --prod` with no auth credentials exits 1.

## Expected behavior

`handleMigrateCommand` and `handleAggregateCommand` should mirror the `handleDeployCommand` pattern and pass `env: getConvexDeploymentCommandEnv()` (gated on `backend === "convex"`) into every downstream `runMigrationFlow` / `runAggregateBackfillFlow` / `runAggregatePruneFlow` call.

Concretely, every subcommand branch that currently reads like:

```js
return runMigrationFlow({ execaFn, backendAdapter, migrationConfig, targetArgs, context: ..., direction: ... });
```

should instead read:

```js
const commandEnv = backend === "convex" ? getConvexDeploymentCommandEnv() : void 0;
return runMigrationFlow({ execaFn, backendAdapter, migrationConfig, targetArgs, env: commandEnv, context: ..., direction: ... });
```

Same treatment for all three `runAggregate*` calls inside `handleAggregateCommand`, and for the `status` / `cancel` branches of `handleMigrateCommand` that go through `runBackendFunction` directly.

## Workaround (until fixed)

Forward the key via `--env-file` instead of ambient env — the flag is passed through to `convex run`, which reads the file directly and bypasses `createBackendCommandEnv`'s wipe:

```yaml
- name: Run Convex migrations (prod only)
  working-directory: packages/backend
  env:
    CONVEX_DEPLOY_KEY: ${{ secrets.CONVEX_DEPLOY_KEY }}
  run: |
    trap 'rm -f "$RUNNER_TEMP/convex-migrate.env"' EXIT
    printf 'CONVEX_DEPLOY_KEY=%s\n' "$CONVEX_DEPLOY_KEY" > "$RUNNER_TEMP/convex-migrate.env"
    bunx kitcn migrate up --yes --env-file "$RUNNER_TEMP/convex-migrate.env"
```

This is the same workaround documented in the pre-#212 release notes for #208 / #209, reintroduced for the standalone `migrate` / `aggregate` paths.

## Notes for maintainers

- `dev` and `codegen` correctly keep clearing ambient deployment env per #212's design; this request is scoped to prod-target subcommands (`deploy`, `migrate`, `aggregate`).
- The "1-second exit with no output" failure mode is painful to debug in CI because `echoOutput: false` hides the convex-side auth error. Consider echoing subprocess stderr on non-zero exit even when `echoOutput` is false, so the underlying "Could not authenticate to Convex" message reaches the log.

--- a/.changeset/serious-pans-sparkle.md
+++ b/.changeset/serious-pans-sparkle.md
@@ -1,0 +1,8 @@
+---
+"kitcn": patch
+---
+
+## Patches
+
+- Fix `kitcn migrate` and `kitcn aggregate` so Convex prod-target runs keep
+  ambient deployment auth env in CI.

--- a/packages/kitcn/src/cli/backend-core.ts
+++ b/packages/kitcn/src/cli/backend-core.ts
@@ -5594,8 +5594,9 @@ export async function runAggregatePruneFlow(params: {
   execaFn: typeof execa;
   backendAdapter: BackendAdapter;
   targetArgs: string[];
+  env?: Record<string, string | undefined>;
 }): Promise<number> {
-  const { execaFn, backendAdapter, targetArgs } = params;
+  const { execaFn, backendAdapter, targetArgs, env } = params;
   const result = await runBackendFunction(
     execaFn,
     backendAdapter,
@@ -5606,6 +5607,7 @@ export async function runAggregatePruneFlow(params: {
     targetArgs,
     {
       echoOutput: false,
+      env,
     }
   );
 

--- a/packages/kitcn/src/cli/commands/aggregate.test.ts
+++ b/packages/kitcn/src/cli/commands/aggregate.test.ts
@@ -1,0 +1,101 @@
+import { describe, expect, mock, test } from 'bun:test';
+import { createDefaultConfig } from '../test-utils';
+import { handleAggregateCommand } from './aggregate';
+
+describe('cli/commands/aggregate', () => {
+  const withConvexDeployKey = async (
+    deployKey: string,
+    run: () => Promise<void>
+  ) => {
+    const originalDeployKey = process.env.CONVEX_DEPLOY_KEY;
+    process.env.CONVEX_DEPLOY_KEY = deployKey;
+
+    try {
+      await run();
+    } finally {
+      process.env.CONVEX_DEPLOY_KEY = originalDeployKey;
+    }
+  };
+
+  test('handleAggregateCommand(backfill) forwards ambient Convex deployment env for convex backend', async () => {
+    const calls: Record<string, string | undefined>[] = [];
+    const execaStub = mock(
+      async (
+        _cmd: string,
+        _args: string[],
+        options?: { env?: Record<string, string | undefined> }
+      ) => {
+        calls.push(options?.env ?? {});
+        return {
+          exitCode: 0,
+          stdout: `${JSON.stringify({ scheduled: 0, targets: 0 })}\n`,
+          stderr: '',
+        } as any;
+      }
+    );
+    const loadConfigStub = mock(() => {
+      const config = createDefaultConfig();
+      config.deploy.aggregateBackfill.wait = false;
+      return config;
+    });
+
+    await withConvexDeployKey('prod:demo|secret', async () => {
+      const exitCode = await handleAggregateCommand(
+        ['aggregate', 'backfill', '--prod'],
+        {
+          realConvex: '/fake/convex/main.js',
+          execa: execaStub as any,
+          loadCliConfig: loadConfigStub as any,
+        }
+      );
+
+      expect(exitCode).toBe(0);
+    });
+
+    expect(calls).toHaveLength(1);
+    expect(calls[0]).toEqual(
+      expect.objectContaining({
+        CONVEX_DEPLOY_KEY: 'prod:demo|secret',
+      })
+    );
+  });
+
+  test('handleAggregateCommand(prune) forwards ambient Convex deployment env for convex backend', async () => {
+    const calls: Record<string, string | undefined>[] = [];
+    const execaStub = mock(
+      async (
+        _cmd: string,
+        _args: string[],
+        options?: { env?: Record<string, string | undefined> }
+      ) => {
+        calls.push(options?.env ?? {});
+        return {
+          exitCode: 0,
+          stdout: `${JSON.stringify({ pruned: 0 })}\n`,
+          stderr: '',
+        } as any;
+      }
+    );
+    const loadConfigStub = mock(() => createDefaultConfig());
+
+    await withConvexDeployKey('prod:demo|secret', async () => {
+      const exitCode = await handleAggregateCommand(
+        ['aggregate', 'prune', '--prod'],
+        {
+          realConvex: '/fake/convex/main.js',
+          execa: execaStub as any,
+          loadCliConfig: loadConfigStub as any,
+        }
+      );
+
+      expect(exitCode).toBe(0);
+    });
+
+    expect(calls).toHaveLength(1);
+    expect(calls[0]).toEqual(
+      expect.objectContaining({
+        CONVEX_DEPLOY_KEY: 'prod:demo|secret',
+      })
+    );
+  });
+});

--- a/packages/kitcn/src/cli/commands/aggregate.ts
+++ b/packages/kitcn/src/cli/commands/aggregate.ts
@@ -2,6 +2,7 @@ import {
   createBackendAdapter,
   extractBackendRunTargetArgs,
   extractBackfillCliOptions,
+  getConvexDeploymentCommandEnv,
   parseArgs,
   type RunDeps,
   resolveBackfillConfig,
@@ -43,6 +44,8 @@ export const handleAggregateCommand = async (
     realConvexPath,
     realConcavePath,
   });
+  const commandEnv =
+    backend === 'convex' ? getConvexDeploymentCommandEnv() : undefined;
   const {
     remainingArgs: aggregateCommandArgs,
     overrides: aggregateBackfillOverrides,
@@ -62,6 +65,7 @@ export const handleAggregateCommand = async (
       execaFn,
       backendAdapter,
       targetArgs,
+      env: commandEnv,
     });
   }
 
@@ -71,6 +75,7 @@ export const handleAggregateCommand = async (
     backfillConfig,
     mode: subcommand === 'rebuild' ? 'rebuild' : 'resume',
     targetArgs,
+    env: commandEnv,
     context: 'aggregate',
   });
 };

--- a/packages/kitcn/src/cli/commands/migrate.test.ts
+++ b/packages/kitcn/src/cli/commands/migrate.test.ts
@@ -13,6 +13,20 @@ import {
 } from './migrate';
 
 describe('cli/commands/migrate', () => {
+  const withConvexDeployKey = async (
+    deployKey: string,
+    run: () => Promise<void>
+  ) => {
+    const originalDeployKey = process.env.CONVEX_DEPLOY_KEY;
+    process.env.CONVEX_DEPLOY_KEY = deployKey;
+
+    try {
+      await run();
+    } finally {
+      process.env.CONVEX_DEPLOY_KEY = originalDeployKey;
+    }
+  };
+
   test('parseMigrateCommandArgs parses create/list/up/down/status/cancel shapes', () => {
     expect(parseMigrateCommandArgs(['create', 'Add', 'field'])).toEqual({
       subcommand: 'create',
@@ -253,5 +267,129 @@ describe('cli/commands/migrate', () => {
     expect(exitCode).toBe(0);
     expect(calls).toHaveLength(2);
     expect(calls[1]?.args).toContain('generated/server:migrationStatus');
+  });
+
+  test('handleMigrateCommand(up) forwards ambient Convex deployment env for convex backend', async () => {
+    const calls: Array<{
+      args: string[];
+      env: Record<string, string | undefined>;
+    }> = [];
+    const execaStub = mock(
+      async (_cmd: string, args: string[], options?: { env?: unknown }) => {
+        calls.push({
+          args,
+          env: (options?.env ?? {}) as Record<string, string | undefined>,
+        });
+        return {
+          exitCode: 0,
+          stdout: `${JSON.stringify({ status: 'noop' })}\n`,
+          stderr: '',
+        } as any;
+      }
+    );
+    const loadConfigStub = mock(() => {
+      const config = createDefaultConfig();
+      config.deploy.migrations.wait = false;
+      return config;
+    });
+
+    await withConvexDeployKey('prod:demo|secret', async () => {
+      const exitCode = await handleMigrateCommand(
+        ['migrate', 'up', '--prod', '--yes'],
+        {
+          realConvex: '/fake/convex/main.js',
+          execa: execaStub as any,
+          loadCliConfig: loadConfigStub as any,
+        }
+      );
+
+      expect(exitCode).toBe(0);
+    });
+
+    expect(calls).toHaveLength(1);
+    expect(calls[0]?.args).toContain('--prod');
+    expect(calls[0]?.env).toEqual(
+      expect.objectContaining({
+        CONVEX_DEPLOY_KEY: 'prod:demo|secret',
+      })
+    );
+  });
+
+  test('handleMigrateCommand(status) forwards ambient Convex deployment env for convex backend', async () => {
+    const calls: Record<string, string | undefined>[] = [];
+    const execaStub = mock(
+      async (
+        _cmd: string,
+        _args: string[],
+        options?: { env?: Record<string, string | undefined> }
+      ) => {
+        calls.push(options?.env ?? {});
+        return {
+          exitCode: 0,
+          stdout: '{}\n',
+          stderr: '',
+        } as any;
+      }
+    );
+    const loadConfigStub = mock(() => createDefaultConfig());
+
+    await withConvexDeployKey('prod:demo|secret', async () => {
+      const exitCode = await handleMigrateCommand(
+        ['migrate', 'status', '--prod'],
+        {
+          realConvex: '/fake/convex/main.js',
+          execa: execaStub as any,
+          loadCliConfig: loadConfigStub as any,
+        }
+      );
+
+      expect(exitCode).toBe(0);
+    });
+
+    expect(calls).toHaveLength(1);
+    expect(calls[0]).toEqual(
+      expect.objectContaining({
+        CONVEX_DEPLOY_KEY: 'prod:demo|secret',
+      })
+    );
+  });
+
+  test('handleMigrateCommand(cancel) forwards ambient Convex deployment env for convex backend', async () => {
+    const calls: Record<string, string | undefined>[] = [];
+    const execaStub = mock(
+      async (
+        _cmd: string,
+        _args: string[],
+        options?: { env?: Record<string, string | undefined> }
+      ) => {
+        calls.push(options?.env ?? {});
+        return {
+          exitCode: 0,
+          stdout: '{}\n',
+          stderr: '',
+        } as any;
+      }
+    );
+    const loadConfigStub = mock(() => createDefaultConfig());
+
+    await withConvexDeployKey('prod:demo|secret', async () => {
+      const exitCode = await handleMigrateCommand(
+        ['migrate', 'cancel', '--prod', '--run-id', 'mr_123'],
+        {
+          realConvex: '/fake/convex/main.js',
+          execa: execaStub as any,
+          loadCliConfig: loadConfigStub as any,
+        }
+      );
+
+      expect(exitCode).toBe(0);
+    });
+
+    expect(calls).toHaveLength(1);
+    expect(calls[0]).toEqual(
+      expect.objectContaining({
+        CONVEX_DEPLOY_KEY: 'prod:demo|secret',
+      })
+    );
   });
 });

--- a/packages/kitcn/src/cli/commands/migrate.ts
+++ b/packages/kitcn/src/cli/commands/migrate.ts
@@ -3,6 +3,7 @@ import {
   extractBackendRunTargetArgs,
   extractMigrationCliOptions,
   extractMigrationDownOptions,
+  getConvexDeploymentCommandEnv,
   parseArgs,
   type RunDeps,
   resolveConfiguredBackend,
@@ -110,6 +111,8 @@ export const handleMigrateCommand = async (
     realConvexPath,
     realConcavePath,
   });
+  const commandEnv =
+    backend === 'convex' ? getConvexDeploymentCommandEnv() : undefined;
 
   if (migrateArgs.subcommand === 'create') {
     const rawName = migrateArgs.restArgs.join(' ').trim();
@@ -142,6 +145,7 @@ export const handleMigrateCommand = async (
       backendAdapter,
       migrationConfig,
       targetArgs,
+      env: commandEnv,
       context: 'migration',
       direction: 'up',
     });
@@ -156,6 +160,7 @@ export const handleMigrateCommand = async (
       backendAdapter,
       migrationConfig,
       targetArgs: downTargetArgs,
+      env: commandEnv,
       context: 'migration',
       direction: 'down',
       steps,
@@ -169,7 +174,10 @@ export const handleMigrateCommand = async (
       backendAdapter,
       'generated/server:migrationStatus',
       {},
-      targetArgs
+      targetArgs,
+      {
+        env: commandEnv,
+      }
     );
     return statusResult.exitCode;
   }
@@ -203,7 +211,10 @@ export const handleMigrateCommand = async (
     backendAdapter,
     'generated/server:migrationCancel',
     runId ? { runId } : {},
-    cancelTargetArgs
+    cancelTargetArgs,
+    {
+      env: commandEnv,
+    }
   );
   return cancelResult.exitCode;
 };

